module/data/item/templates/activities.mjs

```diff
@@ -32,7 +33,58 @@ export default class ActivitiesTemplate extends SystemDataModel {
    * @param {object} source  Candidate source data to migrate.
    */
   static migrateActivities(source) {
-    ActivitiesTemplate.#migrateUses(source);
+    if ( source.system ) ActivitiesTemplate.#migrateUses(source.system);
+    // TODO: Handle migration of data kept on spells and inferred by activities
+    // TODO: Handle migration of base damage, range, & ammunition on weapons
+    if ( this.#shouldMigrateActivity(source) ) this.#migrateInitialActivity(source);
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Method to determine whether the activity creation migration should be performed. This migration should only be
+   * performed on whole item data rather than partial updates, so check to ensure all of the necessary data is present.
+   * @param {object} source  Candidate source data to migrate.
+   * @returns {boolean}
+   */
+  static #shouldMigrateActivity(source) {
+    // If source data doesn't have effects or system data, then is only a partial migration
+    if ( !source.effects || !source.system ) return false;
+
+    // If item doesn't have an action type or activation, then it doesn't need an activity
+    if ( !source.system.actionType || !source.system.activation?.type ) return false;
+
+    // If item was updated after `4.0.0`, it shouldn't need the migration
+    if ( foundry.utils.isNewerVersion(source._stats.systemVersion ?? "0.0.0", "3.9.9") ) return false;
```

Comment:
Prefer to use the negation here, rather than a string like `3.9.9`, since version numbers aren't decimals.

```suggestion
    if ( !foundry.utils.isNewerVersion("4.0.0", source._stats.systemVersion ?? "0.0.0") ) return false;
```